contracts/strategies/VariableRewardsStrategy.sol

```diff
@@ -0,0 +1,304 @@
+        return _calculateDepositFee(_amount);
+    }
+
+    function calculateWithdrawFee(uint256 _amount) public view returns (uint256) {
+        return _calculateWithdrawFee(_amount);
+    }
+
+    /**
+     * @notice Approve tokens for use in Strategy
+     * @dev Deprecated; approvals should be handled in context of staking
+     */
+    function setAllowances() public override onlyOwner {
+        revert("setAllowances::deprecated");
+    }
+
+    /**
+     * @notice Deposit tokens to receive receipt tokens
+     * @param _amount Amount of tokens to deposit
+     */
+    function deposit(uint256 _amount) external override {
+        _deposit(msg.sender, _amount);
+    }
+
+    /**
+     * @notice Deposit using Permit
+     * @param _amount Amount of tokens to deposit
+     * @param _deadline The time at which to expire the signature
+     * @param _v The recovery byte of the signature
+     * @param _r Half of the ECDSA signature pair
+     * @param _s Half of the ECDSA signature pair
+     */
+    function depositWithPermit(
+        uint256 _amount,
+        uint256 _deadline,
+        uint8 _v,
+        bytes32 _r,
+        bytes32 _s
+    ) external override {
+        depositToken.permit(msg.sender, address(this), _amount, _deadline, _v, _r, _s);
+        _deposit(msg.sender, _amount);
+    }
+
+    function depositFor(address _account, uint256 _amount) external override {
+        _deposit(_account, _amount);
+    }
+
+    function _deposit(address _account, uint256 _amount) internal {
+        require(DEPOSITS_ENABLED == true, "VariableRewardsStrategy::Deposits disabled");
+        if (MAX_TOKENS_TO_DEPOSIT_WITHOUT_REINVEST > 0) {
+            (, uint256 estimatedTotalReward) = _checkReward();
+            if (estimatedTotalReward > MAX_TOKENS_TO_DEPOSIT_WITHOUT_REINVEST) {
+                _reinvest();
+            }
+        }
+        require(
+            depositToken.transferFrom(msg.sender, address(this), _amount),
+            "VariableRewardsStrategy::Deposit token transfer failed"
+        );
+        uint256 depositFee = _calculateDepositFee(_amount);
+        _mint(_account, getSharesForDepositTokens(_amount.sub(depositFee)));
+        _stakeDepositTokens(_amount);
+        emit Deposit(_account, _amount);
+    }
+
+    function _getDepositFeeBips() internal view virtual returns (uint256) {
+        return 0;
+    }
+
+    function _calculateDepositFee(uint256 _amount) internal view virtual returns (uint256) {
+        uint256 depositFeeBips = _getDepositFeeBips();
+        return _amount.mul(depositFeeBips).div(_bip());
+    }
+
+    function withdraw(uint256 _amount) external override {
+        uint256 depositTokenAmount = getDepositTokensForShares(_amount);
+        require(depositTokenAmount > 0, "VariableRewardsStrategy::Withdraw amount too low");
+        uint256 withdrawAmount = _withdrawFromStakingContract(_amount);
+        uint256 withdrawFee = _calculateWithdrawFee(depositTokenAmount);
```

Comment:
Depending on the specifics, this might double-count the withdrawFee. For example, if `_withdrawFromStakingContract(_amount)` returns the amount withdrawn after the fee has already been deducted. This is an implementation concern -- consider documenting

Reply:
added docs